xds/internal/balancer/outlierdetection/balancer.go

```diff
@@ -105,3 +135,793 @@ func (bb) ParseConfig(s json.RawMessage) (serviceconfig.LoadBalancingConfig, err
+		// do the following:"
+		// "Unset the timer start timestamp."
+		b.timerStartTime = time.Time{}
+		for _, addrInfo := range b.addrs {
+			// "Uneject all currently ejected addresses."
+			if !addrInfo.latestEjectionTimestamp.IsZero() {
+				b.unejectAddress(addrInfo)
+			}
+			// "Reset each address's ejection time multiplier to 0."
+			addrInfo.ejectionTimeMultiplier = 0
+		}
+	}
+	b.mu.Unlock()
+	b.pickerUpdateCh.Put(lbCfg)
+
+	// then pass the address list along to the child policy.
+	b.childMu.Lock()
+	defer b.childMu.Unlock()
+	return b.child.UpdateClientConnState(balancer.ClientConnState{
+		ResolverState:  s.ResolverState,
+		BalancerConfig: b.cfg.ChildPolicy.Config,
+	})
+}
+
+func (b *outlierDetectionBalancer) ResolverError(err error) {
+	if b.child == nil {
+		return
+	}
+	b.childMu.Lock()
+	defer b.childMu.Unlock()
+	b.child.ResolverError(err)
+}
+
+func (b *outlierDetectionBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	scw, ok := b.scWrappers[sc]
+	if !ok {
+		// Shouldn't happen if passed down a SubConnWrapper to child on SubConn
+		// creation.
+		b.logger.Errorf("UpdateSubConnState called with SubConn that has no corresponding SubConnWrapper")
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		delete(b.scWrappers, scw.SubConn)
+	}
+	b.scUpdateCh.Put(&scUpdate{
+		scw:   scw,
+		state: state,
+	})
+}
+
+func (b *outlierDetectionBalancer) Close() {
+	b.closed.Fire()
+	if b.child != nil {
+		b.childMu.Lock()
+		b.child.Close()
+		b.child = nil
+		b.childMu.Unlock()
+	}
+
+	// Any other cleanup needs to happen (subconns, other resources?)
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if b.intervalTimer != nil {
+		b.intervalTimer.Stop()
+	}
+}
+
+func (b *outlierDetectionBalancer) ExitIdle() {
+	if b.child == nil {
+		return
+	}
+	if ei, ok := b.child.(balancer.ExitIdler); ok {
+		b.childMu.Lock()
+		defer b.childMu.Unlock()
+		ei.ExitIdle()
+		return
+	}
+
+	// Fallback for children handled in clusterimpl balancer Removing SubConns
+	// is defined in API and also in graceful switch balancer, but already done
+	// in ClusterImpl. I guess we should do that here?
+}
+
+// wrappedPicker delegates to the child policy's picker, and when the request
+// finishes, it increments the corresponding counter in the map entry referenced
+// by the subConnWrapper that was picked.
+type wrappedPicker struct {
+	childPicker balancer.Picker
+	noopPicker  bool
+}
+
+func (wp *wrappedPicker) Pick(info balancer.PickInfo) (balancer.PickResult, error) {
+	pr, err := wp.childPicker.Pick(info)
+	if err != nil {
+		return balancer.PickResult{}, err
+	}
+
+	done := func(di balancer.DoneInfo) {
+		if !wp.noopPicker {
+			incrementCounter(pr.SubConn, di)
+		}
+		if pr.Done != nil {
+			pr.Done(di)
+		}
+	}
+	scw, ok := pr.SubConn.(*subConnWrapper)
+	if !ok {
+		return balancer.PickResult{
+			SubConn: pr.SubConn,
+			Done:    done,
+		}, nil
+	}
```

Comment:
When can this happen and why is it OK to receive a subConn which is not our wrapper?

Reply:
Then please throw a comment. Otherwise, it makes the reader think that maybe there are valid cases where this can happen. Thanks.